### DISCUSSION TITLE
Predictive back gesture support on Android

### DISCUSSION THREAD

**Discussion by kirillzyusko** [software-mansion/react-native-screens#2540] [Ideas]
Just curious whether you've considered to add [Predictive back gesture](https://developer.android.com/guide/navigation/custom-back/predictive-back-gesture) support on Android? Here is an [example](https://developer.android.com/guide/navigation/custom-back/support-animations) showing which animations can be used when you implement that 👀 

It looks nice and gives a feel of a native application on Android (for example me - I like that there is an official alternative to a back gesture navigation (that is available on iOS)).

**Comment by kirillzyusko** [comment]
Probably worth to tag @WoLewicki @alduzy @tboba since you guys mostly work with Android?..

**Comment by WoLewicki** [comment]
@kkafar is the main maintainer of the library so it's a question to him right now.

**Comment by kkafar** [comment]
I was looking into it back in 2022 https://github.com/facebook/react-native/issues/34529, however the topic died. I've planned to go back to this at some point, but it has never made the top of my todo list. 

Need to stabilize Android bugs with screens v4 and should have capacity to work on this.

**Comment by bulkinav** [comment]
@kkafar hi. Do you have any news about the implementation of this feature?

**Comment by kkafar** [comment]
Yeah, doing predictive back gesture on Fragment level on Android AFAIK requires FragmentManager's back stack usage, which in turn eliminates possibility of synchronously commiting fragment trasanctions & we need that in our implementation ==> highly likely this won't be ever done in react native screens v4.

**Comment by kkafar** [comment]
I have more context now, because I'm researching things for next major & we might try to land that, however it's months away.

**Comment by mozzius** [comment]
Looks like predictive back gesture support was [added in 0.81](https://reactnative.dev/blog/2025/08/12/react-native-0.81) - what implication does this have for rn-screens?

**Comment by Siddhantshelake** [comment]
Hi all,

I’m experiencing an issue on Android 16 — pressing the hardware back button causes the app to exit immediately instead of navigating to the previous screen within the stack.

Why This Happens

Android 16 no longer calls onBackPressed() or dispatches KEYCODE_BACK for apps targeting API level 36+.

React Native still depends on these legacy callbacks (onBackPressed() → ReactInstanceManager.onBackPressed() → JS BackHandler).

As a result, when predictive back is active, React Native never receives the back event, and Android assumes the activity should exit — causing the app to close.

Any sample code, patch, or guidance would be greatly appreciated 🙏

**Reply by LeOndaz** [reply]
Are you sure you're in a stack? I just experienced this and I found I was using a drawer

**Reply by crstnn** [reply]
Worth calling out that with versions of RN <0.81 you will have this problem if you have set `targetSdkVersion = 36` instead of `35`.

**Reply by dongseok0** [reply]
I'm experiencing same behavior but only after the app is resumed with `RN: 0.83.1` and `targetSdkVersion = 36` 🤔 
And confirmed it's not happening with `android:enableOnBackInvokedCallback="false"` ref. https://stackoverflow.com/a/79861863/1853949

### When the app started new
```
[debug] RedeemScreen backpress handler attached
[debug-nav] index=1 routes=Home > Redeem
**( Back button pressed )
[debug-back] RedeemScreen hardwareBackPress received
[debug-nav] index=0 routes=Home
[debug-appstate] active -> background
```

### When the app is resumed 
```
[debug-appstate] active -> background
[debug-appstate] background -> active
[debug] RedeemScreen backpress handler attached
[debug-nav] index=1 routes=Home > Redeem
**( Back button pressed )
[debug-appstate] active -> background

```

**Comment by MahmoudBahar** [comment]
Hello, has this been added yet or any info about it's release?
@kkafar

**Comment by mukundrex** [comment]
Looking for the support since, apple zoom animation in expo gives better navigation experience in ios than android. We need to close this parity.
